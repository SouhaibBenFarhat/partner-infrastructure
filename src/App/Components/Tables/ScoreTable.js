/* eslint-disable no-script-url */

import React, {useState} from 'react';
import Link from '@material-ui/core/Link/index';
import {makeStyles} from '@material-ui/core/styles/index';
import Table from '@material-ui/core/Table/index';
import TableBody from '@material-ui/core/TableBody/index';
import TableCell from '@material-ui/core/TableCell/index';
import TableHead from '@material-ui/core/TableHead/index';
import TableRow from '@material-ui/core/TableRow/index';
import Title from '../Charts/Title';
import PeopleService from "../../Services/PeopleService";

const useStyles = makeStyles(theme => ({
    seeMore: {
        marginTop: theme.spacing(3),
    },
    selectedColumn: {
        fontWeight: 'bold',
        fontSize: 20,
        cursor: 'pointer'
    },
    defaultColumn: {
        cursor: 'pointer'
    }
}));

export default function ScoreTable({peoples = []}) {
    const classes = useStyles();
    const [rowsNumber, setRowsNumber] = useState(20);
    const [orderByColumn, setOrderByColumn] = useState(null);

    return (
        <React.Fragment>
            <Title>Scores listing</Title>
            <Table size="small">
                <TableHead>
                    <TableRow>
                        <TableCell onClick={() => setOrderByColumn('first_name')}>
                            <span
                                className={(orderByColumn === 'first_name') ? classes.selectedColumn : classes.defaultColumn}>
                            First name
                            </span>
                        </TableCell>
                        <TableCell onClick={() => setOrderByColumn('last_name')}>
                            <span
                                className={(orderByColumn === 'last_name') ? classes.selectedColumn : classes.defaultColumn}>
                            Last name
                            </span>
                        </TableCell>
                        <TableCell onClick={() => setOrderByColumn('gender')}>
                            <span
                                className={(orderByColumn === 'gender') ? classes.selectedColumn : classes.defaultColumn}>
                            Gender
                            </span>
                        </TableCell>
                        <TableCell onClick={() => setOrderByColumn('city')}>
                            <span
                                className={(orderByColumn === 'city') ? classes.selectedColumn : classes.defaultColumn}>
                            City
                            </span>
                        </TableCell>
                        <TableCell onClick={() => setOrderByColumn('country')}>
                            <span
                                className={(orderByColumn === 'country') ? classes.selectedColumn : classes.defaultColumn}>
                            Country
                            </span>
                        </TableCell>
                        <TableCell onClick={() => setOrderByColumn('score')} align="right">
                            <span
                                className={(orderByColumn === 'score') ? classes.selectedColumn : classes.defaultColumn}>
                            Score
                            </span>
                        </TableCell>
                    </TableRow>
                </TableHead>
                <TableBody>
                    {
                        new PeopleService(peoples || []).orderBy(orderByColumn).slice(0, rowsNumber).map(row => (
                            <TableRow key={row.id}>
                                <TableCell>{row.first_name || '--'}</TableCell>
                                <TableCell>{row.last_name || '--'}</TableCell>
                                <TableCell>{row.gender || '--'}</TableCell>
                                <TableCell>{row.city || '--'}</TableCell>
                                <TableCell>{row.country || '--'}</TableCell>
                                <TableCell align="right">{row.score || 0}</TableCell>
                            </TableRow>
                        ))
                    }
                </TableBody>
            </Table>
            <div className={classes.seeMore}>
                <Link color="primary" href="#" onClick={() => setRowsNumber(rowsNumber * 2)}>
                    {peoples && rowsNumber >= peoples.length ? '' : 'See more scores'}
                </Link>
            </div>
            <div className={classes.seeMore}>
                <Link color="primary" href="#" onClick={() => setRowsNumber(peoples.length)}>
                    {peoples && rowsNumber >= peoples.length ? '' : 'Show all'}
                </Link>
            </div>
        </React.Fragment>
    );
}
