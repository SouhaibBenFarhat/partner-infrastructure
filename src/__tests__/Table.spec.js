import React from "react";
import {create} from "react-test-renderer";
import ScoreTable from "../App/Components/Tables/ScoreTable";
import TableRow from '@material-ui/core/TableRow/index';
import TableBody from '@material-ui/core/TableBody/index';
import TableCell from '@material-ui/core/TableCell/index';
import TableHead from '@material-ui/core/TableHead/index';
import Faker from 'faker'

const mockData = (size = 10) => {
    const mockData = [];
    for (let i = 0; i < size; i++) {
        mockData.push({
            "id": Faker.random.uuid(),
            "first_name": Faker.name.firstName(),
            "last_name": Faker.name.lastName(),
            "email": Faker.internet.email(),
            "gender": Faker.lorem.slug(),
            "city": Faker.address.city(),
            "country": Faker.address.country(),
            "score": Faker.random.number(),
            "created_at": Faker.date.past()
        })
    }
    return mockData;
};

describe("ScoreTable component", () => {
    test("should render with empty data", () => {
        const scoreTable = create(<ScoreTable peoples={[]}/>);
        expect(scoreTable.toJSON()).toMatchSnapshot();
    });

    test('Rows matches the data', () => {
        const scoreTableInstance = create(<ScoreTable peoples={mockData(10)}/>).root;
        expect(scoreTableInstance.findByType(TableBody).findAllByType(TableRow).length).toBe(mockData(10).length);
    });

    test('Table header match the data', () => {
        const scoreTableInstance = create(<ScoreTable peoples={mockData(10)}/>).root;
        expect(scoreTableInstance.findByType(TableHead).findAllByType(TableCell).length).toBeLessThanOrEqual(Object.keys(mockData(10)[0]).length);
    });

    test('Table should show 20 rows by default', () => {
        const scoreTableInstance = create(<ScoreTable peoples={mockData(1000)}/>).root;
        expect(scoreTableInstance.findByType(TableBody).findAllByType(TableRow).length).toBe(20);
    });
});
