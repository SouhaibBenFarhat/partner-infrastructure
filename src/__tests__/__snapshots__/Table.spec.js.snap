// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreTable component should render with empty data 1`] = `
Array [
  <h2
    className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
  >
    Scores listing
  </h2>,
  <table
    className="MuiTable-root"
  >
    <thead
      className="MuiTableHead-root"
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
          onClick={[Function]}
          scope="col"
        >
          <span
            className="makeStyles-defaultColumn-3"
          >
            First name
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
          onClick={[Function]}
          scope="col"
        >
          <span
            className="makeStyles-defaultColumn-3"
          >
            Last name
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
          onClick={[Function]}
          scope="col"
        >
          <span
            className="makeStyles-defaultColumn-3"
          >
            Gender
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
          onClick={[Function]}
          scope="col"
        >
          <span
            className="makeStyles-defaultColumn-3"
          >
            City
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
          onClick={[Function]}
          scope="col"
        >
          <span
            className="makeStyles-defaultColumn-3"
          >
            Country
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeSmall MuiTableCell-head"
          onClick={[Function]}
          scope="col"
        >
          <span
            className="makeStyles-defaultColumn-3"
          >
            Score
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
    />
  </table>,
  <div
    className="makeStyles-seeMore-1"
  >
    <a
      className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
      href="#"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      
    </a>
  </div>,
  <div>
    <a
      className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
      href="#"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      
    </a>
  </div>,
]
`;
